### ISSUE TITLE
Binding free "member" function with object reference

### ISSUE THREAD

**Issue by THE-FYP** [ThePhD/sol2#603]
Here is what I want to do:
```cpp
class LuaScript {/* some class */};

void someFreeMemberFunction(LuaScript* script, int a1, float a2) {
  // ...
}

// binding
LuaScript* script = new LuaScript();
lua.set_function("some_func", &someFreeMemberFunction, script); // or like"sol::as_upvalue(script)" instead of just "script"

// calling from lua
some_func(1, 2.3) -- the "script" argument is not passed! it is bound with the function
```
Sol provides ability to bind actual member function with a pointer or reference to its class, and you can do something similar to what I have shown with an userdata methods, but not this.

Why I need this?.. Well, yes, I can use lambdas or functors, it is works, but there is one problem - they're both **significantly** increase size of the application binary. File size grows by 886kbytes just for wrapping about 1700 functions with a bound pointer, and I don't find an explanation for that. I'm aware that it is not just a file size, it is actual machine instructions which is absolutely excessive in this case.
My simplest wrapper in the world:
```cpp
// LuaScript method
template<typename R, typename... Args>
decltype(auto) setFunction(const std::string& key, R(*fx)(LuaScript*, Args...)) {
  return _lua.set_function(key, [this, fx](Args... args) -> R {
    return fx(this, args...);
  });
}
```
I could put these functions inside a class and bind them that way, but these functions are not methods of this class, they just need to have access to script (not just a lua state) from which they're was called, so it wouldn't be a good desicion.

The question is, would this feature fit into sol and can I request for its addition? In case if no, maybe you can give me a hint how I can tweak sol for this?

**Comment by ThePhD** [comment]
sol2 doesn't provide an abstraction for upvalues for C Functions, namely because sol2 is the one to abuse upvalues to store and retrieve many of the things it needs to work. I agree it owould be ideal, but it is problematic because "hammering" down what upvalues sol2 uses and then trying to make the rest of the space available to the user would be an... interesting problem, to say the least.

As a side note, is there only 1 "Script" object per state? If so, storing it in the state in some fixed location and then plucking it out again on the other side might help. If not, then I need to bikeshed what a `sol::upvalues` abstraction might look like...

**Comment by THE-FYP** [comment]
To clarify, `sol::as_upvalue` is just the first more or less suitable name that came to my mind, not very descriptive. The issue is not about full support of upvalues, I'm just asking for ability to bind free function as a pseudo-method, don't know how to call it correct.
Possibility to work with upvalues would be nice, but as I suppose it will take a much more effort than that, and lambdas does this job well.

Yes, one LuaScript instance per lua state. Do you mean like storing pointer to LuaScript in a global table? Wouldn't it possibly be broken with a lua environments? Anyway, I don't feel that as a better solution, just a workaround...

**Comment by ThePhD** [comment]
Environments can't hide things from the Lua C API, and subsequently from the sol2 interface. We can say it'd be a workaround all we want, but at the end of the day you need something to work. My suggestion would be to actually [write a customization point](http://sol2.readthedocs.io/en/latest/tutorial/customization.html) for `LuaScript` that would get it from some point in the global table.

For what it's worth, if I were to "add support" for this into sol2 it would essentially be by making a custom struct that essentially behaves exactly like a lambda. You'd have all the same function spam, and a bunch of code that does exactly what you do except in a much more convoluted and generic way (and probably bloat your binary). Another way you can try to solve this is with `std::bind`, which may employ some magic underneath that does this in a much more simply way... not that I have any hope that sol2 actually works with `std::bind`, since that too just does what lambdas do + magic to handle placeholders.

In your case, I would actually recommend writing a customization point. That way, the Lua system will only need +2 structs / +2 functions (3 if you want to write a checker for it too) to always pull the thing from the same place in Lua. My suggestion is to use a string key that is filled with numbers, periods, and non-valid-utf8 sequences to make it next to impossible to "accidentally grab" in a script. sol2 employs this for some of the global storage it allocates for itself, albeit it just uses dots and mildly relevant emoji codepoints.

**Comment by ThePhD** [comment]
The other solution would be for me to make upvalue support, and have people write functions in some fashion such as

````
void f( sol::upvalue<LuaScript>, int param1, my_struct& param2 ) {

}
````

And make it callable from Lua like `my_bound_f( 1, my_object )`

**Comment by THE-FYP** [comment]
>For what it's worth, if I were to "add support" for this into sol2 it would essentially be by making a custom struct that essentially behaves exactly like a lambda. You'd have all the same function spam, and a bunch of code that does exactly what you do except in a much more convoluted and generic way (and probably bloat your binary).

I was looking through sol's source code before opening this issue and it seemed to me that it's possible to implement this the similar way to how binding of member functions is implemented, but I don't know all the complexities of sol's work, so it may not be true.

>Another way you can try to solve this is with std::bind, which may employ some magic underneath that does this in a much more simply way... not that I have any hope that sol2 actually works with std::bind, since that too just does what lambdas do + magic to handle placeholders.

Was trying it too, sol doesn't recognize std::_Binder as a valid functor, but it works when wrapped into std::function, either way it doesn't help because it also leads to code bloat even more and unnecessarily increases memory allocations at runtime.

I don't think implementing the whole upvalue system worth it in this particular case, and again, lambda can do it for you when needed. Anyway, if you think sol needs this feature, I'm not stopping you from doing that. I'll try to use a customization point and let you know how it went.

Thanks for your help

**Comment by ThePhD** [comment]
> I'll try to use a customization point and let you know how it went.

I usually include an implementation of these things with my posts, but I was tired! Here's a way you can write the customization point, to always pull/push to some global area in memory.

````cpp
constexpr static const auto& script_key = "LuaScript.SuperIdentifierWee"; 

namespace sol {
     template <>
     struct getter<LuaScript*> {
          static LuaScript* get(lua_State* L, int index, record& tracking) {
               tracking.use(0); // not actually pulling anything off the stack
               (void)index; // unused
               stack::get_field<true>(L, script_key);
               LuaScript* ls = static_cast<LuaScript*>(lua_tolightuserdata(L, -1);
               lua_pop(L, 1); // clean up stack value returned by `get_field`
               return ls;
     };
     
     template <>
     struct pusher<LuaScript*> {
          static int push(lua_State* L, LuaScript* ls) {
               // push light userdata into global table
               stack::set_field<true>(L, script_key, make_light(ls));
               // return nothing on the stack
               return 0;
     };
}
````

> I was looking through sol's source code before opening this issue and it seemed to me that it's possible to implement this the similar way to how binding of member functions is implemented, but I don't know all the complexities of sol's work, so it may not be true.

The way it's done is with [this piece of code here](https://github.com/ThePhD/sol2/blob/develop/sol/function_types_stateful.hpp#L57). We create a template which knows the object type `T`, which is more or less derived from what you pass in. Note that this is only a single parameter... _and_ that it's entirely separate from its sister class in the `function_types_stateless.hpp`. In order you implement what you're suggesting, we'd make a parameter pack that tells us the types of the functions we're tying to the function at bound time. PLUS, we'd need to support overriding not just the 1st argument, but the 1st and 3rd, or just the 2nd. So we'd need to attach a compile-time integer to those too, or some form of placeholder value to indicate what we are / are not binding.

At this point, we have a templated struct that takes 2+ things: the function type, the member if necessary, and a list of things to bind. Then, internally we create 2 or 3 static functions to properly trampoline + call this function, so that's 2 or 3 additional functions per function-type-specific, binding-type-specific call we make...

In other words, we're approaching how much function spam we're hitting for the lambda. Not to mention we have to *explicitly* unpack all of this, right? So we need to write a custom evaluator that pulls out arguments from some kind of storage: probably another `std::tuple`, which already makes our compile times BLARGH-worthy, and...

... Can you see how this just makes your problem worse?

**Comment by THE-FYP** [comment]
I already made my own, working fine, now tested yours, there is two problems: checker is necessary otherwise sol won't be happy with that; by some reason `LuaScript* s ...; lua.set(script_key, s);` doesn't work as expected, you need to do `lua.set(script_key, sol::make_light(s));` to get it behave normal, but this way custom `pusher::push` is not called. Except these two things it is also working fine.

Here is mine:
```cpp
class ThisScript : public LuaScript { ThisScript() = delete; };

constexpr const char* THIS_SCRIPT_GLOBAL_KEY = ".this_script";

namespace sol {
	template <>
	struct lua_type_of<ThisScript*> : std::integral_constant<sol::type, sol::type::poly> {};

	namespace stack {
		template <>
		struct checker<ThisScript*> {
			template <typename Handler>
			static bool check(lua_State* L, int index, Handler&& handler, record& tracking) {
				return true;
			}
		};

		template <>
		struct getter<ThisScript*> {
			static ThisScript* get(lua_State* L, int index, record& tracking) {
				lua_getglobal(L, THIS_SCRIPT_GLOBAL_KEY);
				return static_cast<ThisScript*>(stack::pop<LuaScript*>(L));
			}
		};
	}
}

...

lua.set(THIS_SCRIPT_GLOBAL_KEY, new LuaScript());
// and use ThisScript* in function arguments whenever you want to get it
```
>The way it's done is with this piece of code here. We create a template which knows the object type T, which is more or less derived from what you pass in. Note that this is only a single parameter... and that it's entirely separate from its sister class in the function_types_stateless.hpp.

That's where my mistake was, I thought `upvalue_member_function` is responsible for binding class method with an instance pointer and was looking from this even trying to tweak sol. Derp
Okay, now I see the reasons, and by the way, when `upvalue_member_function` is used then?

**Comment by ThePhD** [comment]
`upvalue_member_function` is used when the binding class meets the criteria of the passed-in `T` being a reference of some sort (`T*` or `std::reference_wrapper<T>`). Therefore, it stores the reference-like `T` as a pointer in an upvalue (as a light userdata, in particular).

Also, that's a memory leak. Using `lua.set(THIS_SCRIPT_GLOBAL_KEY, new LuaScript());` will not call the destructor. This is talked about here: http://sol2.readthedocs.io/en/latest/tutorial/ownership.html

Use `lua.set(THIS_SCRIPT_GLOBAL_KEY, std::make_unique<LuaScript>());` instead.

Sorry about the checker: I forgot it's absolutely necessary (it's for overload resolution and type checking). In order to use my approach you also have to define the type to be `lightuserdata` for `lua_type_of`.

using `lua_getglobal` instead of the `stack::get_field` stuff is also fine too, since that's essentially what it boils down to.

**Comment by ThePhD** [comment]
HEre is a more full, complete example that I actually compiled to make sure I'm not tripping:

````
constexpr static const auto& script_key = "LuaScript.SuperIdentifierWee";

struct LuaScript {
	int value = 2;
};

namespace sol {
	template <>
	struct lua_type_of<LuaScript*> : std::integral_constant<sol::type, sol::type::lightuserdata> {};

	namespace stack {
		template <>
		struct checker<LuaScript*> {
			template <typename Handler>
			static bool check(lua_State* L, int index, Handler&& handler, record& tracking) {
				stack::get_field<true>(L, script_key);
				type t = static_cast<type>(lua_type(L, -1));
				lua_pop(L, 1);
				return t == type::lightuserdata;
			}
		};

		template <>
		struct getter<LuaScript*> {
			static LuaScript* get(lua_State* L, int index, record& tracking) {
				tracking.use(0); // not actually pulling anything off the stack
				(void)index; // unused
				stack::get_field<true>(L, script_key);
				LuaScript* ls = static_cast<LuaScript*>(lua_touserdata(L, -1));
				lua_pop(L, 1); // clean up stack value returned by `get_field`
				return ls;
			};
		};
		template <>
		struct pusher<LuaScript*> {
			static int push(lua_State* L, LuaScript* ls) {
				// push light userdata
				return stack::push(L, make_light(ls));;
			}
		};
	}
}

int main() {
	
	sol::state lua;
	lua.open_libraries(sol::lib::base);

	LuaScript instance;

	lua.set_function("f", [](LuaScript* l, int value) {
		return l->value + value;
	});
	lua.set(script_key, &instance);

	lua.script("assert(f(1) == 3)");

	return 0;
}
````

I'm going to leave this issue open to track maybe doing something like `sol::upvalues` in the near future...

**Comment by THE-FYP** [comment]
>Also, that's a memory leak. Using lua.set(THIS_SCRIPT_GLOBAL_KEY, new LuaScript()); will not call the destructor. This is talked about here: http://sol2.readthedocs.io/en/latest/tutorial/ownership.html

Yep, I know it, not new to C++ and sol :) It was just a quick example for testing.

I already switched my project to using this customization point, no bloating of binary anymore and everything is working smoothly. And now I changed my mind about using this method, it has its own advantage - you can get pointer to LuaScript from anywhere you want without binding anything explicitly, and it is very handy.

Thank you

**Comment by Alterise** [comment]
> I'm actually going to tombstone this for a bit. I've got some other things I need to work on. If someone else searches this and they see this, drop a note here and I'll make sure to put it on the High Priority list.

*Dropped*
